packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,416 @@
+ * @param {stream} callback.connection - A duplex stream.
+ */
+ConnectionPool.prototype.acquire = function(id, callback) {
+  var self = this;
+
+  if (is.fn(id)) {
+    callback = id;
+    id = null;
+  }
+
+  if (!this.isOpen) {
+    callback(new Error('No connections available to make request.'));
+    return;
+  }
+
+  // it's possible that by the time a user acks the connection could have
+  // closed, so in that case we'll just return any connection
+  if (!this.connections.has(id)) {
+    id = getFirstConnectionId();
+  }
+
+  var connection = this.connections.get(id);
+
+  if (connection) {
+    callback(null, connection);
+    return;
+  }
+
+  this.once('connected', function(connection) {
+    callback(null, connection);
+  });
+
+  function getFirstConnectionId() {
+    return self.connections.keys().next().value;
+  }
+};
+
+/**
+ * Ends each connection in the pool and closes the pool, preventing new
+ * connections from being created.
+ *
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.error - An error returned while closing the pool.
+ */
+ConnectionPool.prototype.close = function(callback) {
+  var self = this;
+  var connections = Array.from(this.connections.values());
+
+  callback = callback || common.util.noop;
+
+  this.isOpen = false;
+  self.connections.clear();
+
+  this.queue.forEach(clearTimeout);
+  this.queue.length = 0;
+
+  each(connections, function(connection, onEndCallback) {
+    connection.end(onEndCallback);
+  }, callback);
+};
+
+/**
+ * Creates a connection. This is async but instead of providing a callback
+ * a `connected` event will fire once the connection is ready.
+ */
+ConnectionPool.prototype.createConnection = function() {
+  var self = this;
+
+  this.getClient(function(err, client) {
+    if (err) {
+      self.emit('error', err);
+      return;
+    }
+
+    var id = uuid.v4();
+    var connection = client.streamingPull(self.metadata_);
+    var errorImmediateHandle;
+
+    connection.on('error', function(err) {
+      // since this is a bidi stream it's possible that we recieve errors from
+      // reads or writes. We also want to try and cut down on the number of
+      // errors that we emit if other connections are still open. So by using
+      // setImmediate we're able to cancel the error message if it gets passed
+      // to the `status` event where we can check if the connection should be
+      // re-opened or if we should send the error to the user
+      errorImmediateHandle = setImmediate(self.emit.bind(self), 'error', err);
+    });
+
+    connection.on('metadata', function(metadata) {
+      if (!metadata.get('date').length) {
+        return;
+      }
+
+      connection.isConnected = true;
+      self.noConnectionsTime = 0;
+      self.failedConnectionAttempts = 0;
+      self.emit('connected', connection);
+    });
+
+    connection.on('status', function(status) {
+      clearImmediate(errorImmediateHandle);
+
+      connection.end();
```

Comment:
Does a status event always mean the connection should be ended? Are there benign status events?